### PR TITLE
Add Inventory Chunk Download Lambda Skeleton plus unit tests

### PR DESCRIPTION
**Pull request by dbb-aws-only**
Add the skeleton for Inventory Chunk Download Lambda and add unit test cases.

### CODE REVIEW

File: source/refreezer/infrastructure/stack.py [hunk 1]
```diff
@@ -233,6 +260,33 @@ def __init__(
...
+        NagSuppressions.add_resource_suppressions(
+            inventory_retrieval_state_machine.role.node.find_child(
+                "DefaultPolicy"
+            ).node.find_child(
+                "Resource"
+            ),  #  .role.node.find_child("Resource"),
+            [
+                {
+                    "id": "AwsSolutions-IAM5",
+                    "reason": "CDK grants AWS managed policy for Lambda basic execution by default. Replacing it with a customer managed policy will be addressed later.",
```
Comment (bassemwanis) [inline source/refreezer/infrastructure/stack.py:280]: The reason needs to address AwsSolutions-IAM5, explaining why the IAM entity contains wildcard permissions.

File: source/refreezer/infrastructure/stack.py [hunk 2]
```diff
@@ -189,11 +192,35 @@ def __init__(
...
+                    ],
+                }
+            ],
+        )
+
+        # Add lambda invoke step
+        inventory_chunk_download_lambda = tasks.LambdaInvoke(
             self,
             "InventoryChunkDownloadLambda",
-            parameters={"InventoryRetrieved": "TRUE"},
```
Comment (bassemwanis) [inline source/refreezer/infrastructure/stack.py:196]: Did removing this line break the pipeline integration test?

File: source/refreezer/infrastructure/stack.py [hunk 3]
```diff
@@ -189,11 +192,35 @@ def __init__(
             ],
         )
 
-        # TODO: To be replaced by InventoryChunkDownload LambdaInvoke task
-        inventory_chunk_download_lambda = sfn.Pass(
+        # create an empty lambda function to pass into the invoke
```
Comment (bassemwanis) [inline source/refreezer/infrastructure/stack.py:195]: No need to have this comment

File: source/refreezer/infrastructure/stack.py [hunk 4]
```diff
@@ -233,6 +260,33 @@ def __init__(
...
+        inventory_chunk_download_lambda_function_logical_id = Stack.of(
+            self
+        ).get_logical_id(inventory_chunk_download_lambda_function.node.default_child)
+
+        NagSuppressions.add_resource_suppressions(
+            inventory_retrieval_state_machine.role.node.find_child(
+                "DefaultPolicy"
+            ).node.find_child(
+                "Resource"
+            ),  #  .role.node.find_child("Resource"),
```
Comment (bassemwanis) [inline source/refreezer/infrastructure/stack.py:276]: No need to have this comment

**Review by bassemwanis — CHANGES_REQUESTED** [review]
Can you squash your commits into 1 commit?

File: source/refreezer/infrastructure/stack.py [hunk 5]
```diff
@@ -189,11 +192,42 @@ def __init__(
...
+        )
+
+        # Add lambda invoke step
+        inventory_chunk_download_lambda = tasks.LambdaInvoke(
             self,
             "InventoryChunkDownloadLambda",
-            parameters={"InventoryRetrieved": "TRUE"},
+            lambda_function=inventory_chunk_download_lambda_function,
+        )
+        # result_selector={"InventoryRetrieved": "TRUE"},
```
Comment (bassemwanis) [inline source/refreezer/infrastructure/stack.py:225]: Please remove this comment.

File: source/tests/unit/infrastructure/test_stack.py [hunk 6]
```diff
@@ -179,28 +179,52 @@ def test_get_inventory_step_function_created(
...
+                                    r'"Iterator":{"StartAt":"InventoryChunkDownloadLambda","States":'
+                                    r'{"InventoryChunkDownloadLambda":{"End":true,"Retry":\[{"ErrorEquals":\["Lambda.ServiceException","Lambda.AWSLambdaException","Lambda.SdkClientException"\],'
+                                    r'"IntervalSeconds":\d,"MaxAttempts":\d,"BackoffRate":\d}\],"Type":"Task",'
+                                )
+                            ),
+                            {"Ref": "AWS::Partition"},
+                            assertions.Match.string_like_regexp(
+                                (
+                                    r'states:::lambda:invoke",'
+                                    r'"Parameters":{"FunctionName"'
```
Comment (bassemwanis) [inline source/tests/unit/infrastructure/test_stack.py:213]: There is a missing `:"` at the end of this line
`r'"Parameters":{"FunctionName":"'`

File: source/tests/unit/infrastructure/test_stack.py [hunk 7]
```diff
@@ -179,28 +179,52 @@ def test_get_inventory_step_function_created(
...
+                                )
+                            ),
+                            {"Fn::GetAtt": [inventory_lambda_logical_id, "Arn"]},
+                            assertions.Match.string_like_regexp(
+                                (
+                                    r'"Payload.\$":"\$"}}}},'
```
Comment (bassemwanis) [inline source/tests/unit/infrastructure/test_stack.py:219]: There is a missing `",` at this line
`r'","Payload.\$":"\$"}},'`

File: source/tests/unit/infrastructure/test_stack.py [hunk 8]
```diff
@@ -179,28 +179,52 @@ def test_get_inventory_step_function_created(
...
+                                    r'"States":{"Provided Inventory\?":{"Type":"Choice","Choices":\['
+                                    r'{"Variable":"\$.provided_inventory","StringEquals":"YES","Next":"GlueOrderArchives"}\],'
+                                    r'"Default":"InitiateJob"},'
+                                    r'"InitiateJob":{"Type":"Pass","Next":"DynamoDBPut"},'
+                                    r'"DynamoDBPut":{"Type":"Pass","Next":"GenerateChunkArrayLambda"},'
+                                    r'"GenerateChunkArrayLambda":{"Type":"Pass","Parameters":{"chunk_array":\["\d+-\d+"(,+"\d+-\d+")+\]},"Next":"DistributedMap"},'
+                                    r'"DistributedMap":{"Type":"Map","Next":"GlueOrderArchives",'
+                                    r'"Iterator":{"StartAt":"InventoryChunkDownloadLambda","States":'
+                                    r'{"InventoryChunkDownloadLambda":{"End":true,"Retry":\[{"ErrorEquals":\["Lambda.ServiceException","Lambda.AWSLambdaException","Lambda.SdkClientException"\],'
+                                    r'"IntervalSeconds":\d,"MaxAttempts":\d,"BackoffRate":\d}\],"Type":"Task",'
```
Comment (bassemwanis) [inline source/tests/unit/infrastructure/test_stack.py:206]: There is a missing `"Resource":"arn:` at the end of this line
`r'"IntervalSeconds":\d,"MaxAttempts":\d,"BackoffRate":\d}\],"Type":"Task","Resource":"arn:'`

**Review by bassemwanis — CHANGES_REQUESTED** [review]

File: source/refreezer/infrastructure/stack.py [hunk 9]
```diff
@@ -189,11 +192,35 @@ def __init__(
...
+        inventory_chunk_download_lambda_function = lambda_.Function(
+            self,
+            "InventoryChunkDownload",
+            handler="refreezer.application.handlers.inventory_chunk_download_lambda_handler",
+            runtime=lambda_.Runtime.PYTHON_3_9,
+            code=lambda_.Code.from_asset("source"),
+            description="Lambda to download inventory chunks from Glacier.",
+        )
+
+        assert inventory_chunk_download_lambda_function.role is not None
```
[the remark this answers is not in the document]
Comment (dbb-aws-only) [reply source/refreezer/infrastructure/stack.py:205]: The assert is only there for tox -e type

File: source/refreezer/infrastructure/stack.py (hunk 2 above)
Comment (dbb-aws-only) [reply source/refreezer/infrastructure/stack.py:196]: It did - as you suggested, I added the return value to the lambda handler.

File: source/refreezer/infrastructure/stack.py (hunk 3 above)
Comment (dbb-aws-only) [reply source/refreezer/infrastructure/stack.py:195]: Done.

File: source/refreezer/infrastructure/stack.py (hunk 1 above)
Comment (dbb-aws-only) [reply source/refreezer/infrastructure/stack.py:287]: Done.

File: source/refreezer/infrastructure/stack.py [hunk 10]
```diff
@@ -233,6 +266,31 @@ def __init__(
...
+                "DefaultPolicy"
+            ).node.find_child("Resource"),
+            [
+                {
+                    "id": "AwsSolutions-IAM5",
+                    "reason": "By default wildcard permission is granted to the lambda.  This will be replaced with a proper IAM role later.",
+                    "appliesTo": [
+                        "Resource::<"
+                        + inventory_chunk_download_lambda_function_logical_id
+                        + ".Arn>:*"
```
Comment (kamyarz-aws) [inline source/refreezer/infrastructure/stack.py:288]: use f-string instead of + for string concatenation.

Comment (kamyarz-aws) [reply source/refreezer/infrastructure/stack.py:288]: image

**Review by bassemwanis — APPROVED** [review]

**Review by kamyarz-aws — APPROVED** [review]

File: source/refreezer/infrastructure/stack.py (hunk 10 above)
Comment (kamyarz-aws) [reply source/refreezer/infrastructure/stack.py:338]: This comment is not addressed?

File: source/refreezer/infrastructure/stack.py [hunk 11]
```diff
@@ -234,12 +237,44 @@ def __init__(
                 },
             ],
         )
+        # Create the lambda invoke taks with the suppression, (no output), unit test
+        # we need to have a mechanism to unit test Lambda - Simon, Bassem & Darren - tomorrow
```
Comment (bassemwanis) [inline source/refreezer/infrastructure/stack.py:242]: Please remove these 2 lines

File: source/tests/unit/infrastructure/test_stack.py [hunk 12]
```diff
@@ -228,6 +230,52 @@ def test_get_inventory_step_function_created(
...
+    #                     r'"InitiateJob":{"Type":"Pass","Next":"DynamoDBPut"},'
+    #                     r'"DynamoDBPut":{"Type":"Pass","Next":"GenerateChunkArrayLambda"},'
+    #                     r'"GenerateChunkArrayLambda":{"Type":"Pass","Parameters":{"chunk_array":\["\d+-\d+"(,+"\d+-\d+")+\]},"Next":"DistributedMap"},'
+    #                     r'"DistributedMap":{"Type":"Map","Next":"GlueOrderArchives",'
+    #                     r'"Iterator":{"StartAt":"InventoryChunkDownloadLambda","States":'
+    #                     r'{"InventoryChunkDownloadLambda":{"End":true,"Retry":\[{"ErrorEquals":\["Lambda.ServiceException","Lambda.AWSLambdaException","Lambda.SdkClientException"\],'
+    #                     r'"IntervalSeconds":\d+,"MaxAttempts":\d+,"BackoffRate":\d+}\],"Type":"Task",'
+    #                     r'"Resource":'
+    #                 )
+    #             ),
```
Comment (bassemwanis) [inline source/tests/unit/infrastructure/test_stack.py:277]: Please remove this commented block.

**Review by bassemwanis — CHANGES_REQUESTED** [review]

File: source/refreezer/infrastructure/stack.py (hunk 10 above)
Comment (dbb-aws-only) [reply source/refreezer/infrastructure/stack.py:338]: I did change the source code and it's in this PR - this is the new line of code: 
{"Fn::GetAtt": [inventory_lambda_logical_id, "Arn"]},

File: source/refreezer/infrastructure/stack.py (hunk 11 above)
Comment (dbb-aws-only) [reply source/refreezer/infrastructure/stack.py:240]: Done.  Oops.

File: source/refreezer/infrastructure/stack.py (hunk 10 above)
Comment (bassemwanis) [reply source/refreezer/infrastructure/stack.py:336]: It should be something like that: `f"Resource::<{inventory_chunk_download_lambda_function_logical_id}.Arn>:*"`

**Review by bassemwanis — APPROVED** [review]

**Review by bassemwanis — APPROVED** [review]

**Pull request merged**